File: CRM-Client/src/api/customerFollowUp.ts
/* eslint-disable crmwolf/require-zod-schema */
import request from '@/utils/request'

export interface OwnerInfo {
  id: string
  name: string
  avatar_url: string | null
}

export interface CustomerBasicInfo {
  id: number
  account_name: string
}

export interface CustomerFollowUpCreate {
  method: string
  content: string
  next_follow_time?: string | null
  next_action?: string | null
}

export interface CustomerFollowUpUpdate {
  method?: string | null
  content?: string | null
  next_follow_time?: string | null
  next_action?: string | null
}

export interface CustomerFollowUpResponse {
  id: number
  customer_id: number | null
  original_lead_id: number | null
  content: string
  method: string
  next_follow_time: string | null
  next_action: string | null
  creator_id: string
  creator_info?: OwnerInfo
  customer_info?: CustomerBasicInfo
  created_time: string
  effectiveness_score?: number | null
  effectiveness_is_valid?: boolean | null
  effectiveness_reason?: string | null
  effectiveness_detail_json?: string | null
  effectiveness_status?: 'PENDING' | 'GENERATING' | 'COMPLETED' | 'FAILED' | string | null
  effectiveness_evaluated_time?: string | null
  effectiveness_error_message?: string | null
}

export interface NextFollowUpTimeUpdate {
  next_follow_time: string
}

const customerFollowUpApi = {
  createFollowUp: (customerId: number, data: CustomerFollowUpCreate): Promise<CustomerFollowUpResponse> => {
    return request.post<CustomerFollowUpResponse>(`/v1/customer-follow-ups/${customerId}`, data)
  },

  getFollowUps: (customerId: number): Promise<CustomerFollowUpResponse[]> => {
    return request.get<CustomerFollowUpResponse[]>(`/v1/customer-follow-ups/${customerId}`)
  },

  updateFollowUp: (followUpId: number, data: CustomerFollowUpUpdate): Promise<CustomerFollowUpResponse> => {
    return request.put<CustomerFollowUpResponse>(`/v1/customer-follow-ups/${followUpId}`, data)
  },

  updateNextFollowUpTime: (followUpId: number, data: NextFollowUpTimeUpdate): Promise<CustomerFollowUpResponse> => {
    return request.patch<CustomerFollowUpResponse>(`/v1/customer-follow-ups/${followUpId}/next-time`, data)
  },

  deleteFollowUp: (followUpId: number): Promise<{ message: string }> => {
    return request.delete<{ message: string }>(`/v1/customer-follow-ups/${followUpId}`)
  }
}

export default customerFollowUpApi

File: CRM-Server/app/schemas/customer_follow_up.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional
from datetime import datetime


class OwnerInfo(BaseModel):
    id: str = Field(..., description="用户飞书Open ID")
    name: str = Field(..., description="用户姓名")
    avatar_url: Optional[str] = Field(None, description="用户头像URL")
    
    class Config:
        from_attributes = True


class CustomerBasicInfo(BaseModel):
    id: int = Field(..., description="客户ID")
    account_name: str = Field(..., description="客户公司名称")
    
    class Config:
        from_attributes = True


class CustomerFollowUpBase(BaseModel):
    content: str = Field(..., min_length=1, description="跟进内容")
    method: str = Field(..., min_length=1, max_length=50, description="跟进方式")
    next_follow_time: Optional[datetime] = Field(None, description="计划下次跟进时间")
    next_action: Optional[str] = Field(None, description="下一步动作内容")
    
    @field_validator('content')
    @classmethod
    def content_must_not_be_empty(cls, v):
        if v is not None and (not v or not v.strip()):
            raise ValueError('跟进内容不能为空')
        return v.strip() if v else v
    
    @field_validator('method')
    @classmethod
    def method_must_not_be_empty(cls, v):
        if v is not None and (not v or not v.strip()):
            raise ValueError('跟进方式不能为空')
        return v.strip() if v else v


class CustomerFollowUpCreate(CustomerFollowUpBase):
    pass


class CustomerFollowUpUpdate(BaseModel):
    method: Optional[str] = Field(None, min_length=1, max_length=50, description="跟进方式")
    content: Optional[str] = Field(None, min_length=1, description="跟进内容")
    next_follow_time: Optional[datetime] = Field(None, description="计划下次跟进时间")
    next_action: Optional[str] = Field(None, description="下一步动作内容")

    @field_validator('content')
    @classmethod
    def content_must_not_be_empty(cls, v):
        if v is not None and (not v or not v.strip()):
            raise ValueError('跟进内容不能为空')
        return v.strip() if v else v

    @field_validator('method')
    @classmethod
    def method_must_not_be_empty(cls, v):
        if v is not None and (not v or not v.strip()):
            raise ValueError('跟进方式不能为空')
        return v.strip() if v else v


class CustomerFollowUpResponse(BaseModel):
    id: int = Field(..., description="跟进记录ID")
    customer_id: Optional[int] = Field(None, description="客户ID")
    original_lead_id: Optional[int] = Field(None, description="原始线索ID")
    content: str = Field(..., description="跟进内容")
    method: str = Field(..., description="跟进方式")
    next_follow_time: Optional[datetime] = Field(None, description="计划下次跟进时间")
    next_action: Optional[str] = Field(None, description="下一步动作内容")
    creator_id: str = Field(..., description="创建人飞书用户ID")
    created_time: datetime = Field(..., description="创建时间")
    creator_info: Optional[OwnerInfo] = Field(None, description="跟进人信息")
    customer_info: Optional[CustomerBasicInfo] = Field(None, description="客户基本信息")


class MessageResponse(BaseModel):
    message: str
